File: athru_service/athru_service/utils/hd_ticket.py
"""HD Ticket (Service Request) helpers for Athru Service."""

from __future__ import annotations

import frappe
from frappe import _
from frappe.utils import get_datetime, now_datetime


TICKET_TYPES = [
	"Installation",
	"Breakdown",
	"Preventive",
	"Contract Visit",
	"Billable",
	"Parts",
	"Advisory",
	"Other",
]


def ensure_ticket_types():
	if not frappe.db.exists("DocType", "HD Ticket Type"):
		return
	for name in TICKET_TYPES:
		if frappe.db.exists("HD Ticket Type", name):
			continue
		try:
			frappe.get_doc({"doctype": "HD Ticket Type", "name": name}).insert(ignore_permissions=True)
		except Exception:
			# Some Helpdesk versions use different field layout
			doc = frappe.new_doc("HD Ticket Type")
			doc.name = name
			if hasattr(doc, "description"):
				doc.description = name
			doc.insert(ignore_permissions=True)


def make_service_request_number(when=None) -> str:
	dt = get_datetime(when) if when else now_datetime()
	prefix = dt.strftime("%y%m%d")
	day_start = dt.replace(hour=0, minute=0, second=0, microsecond=0)
	day_end = dt.replace(hour=23, minute=59, second=59, microsecond=0)
	count = frappe.db.count(
		"HD Ticket",
		filters={"creation": ["between", [day_start, day_end]]},
	)
	seq = count + 1
	while True:
		candidate = f"{prefix}-{seq:02d}"
		exists = frappe.db.exists(
			"HD Ticket", {"custom_service_request_number": candidate}
		)
		if not exists:
			return candidate
		seq += 1


def before_insert(doc, method=None):
	if not doc.get("custom_service_request_number"):
		doc.custom_service_request_number = make_service_request_number()
	_stamp_previous_visits(doc)
	_stamp_warranty(doc)


def validate(doc, method=None):
	if not doc.get("custom_service_request_number"):
		doc.custom_service_request_number = make_service_request_number(doc.creation)
	_stamp_previous_visits(doc)
	_stamp_warranty(doc)


def on_update(doc, method=None):
	pass


def _stamp_previous_visits(doc):
	mi = doc.get("custom_machine_installation")
	if not mi:
		return
	doc.custom_total_previous_visits = frappe.db.count(
		"Maintenance Visit", {"custom_machine_installation": mi, "docstatus": ["<", 2]}
	)


def _stamp_warranty(doc):
	mi = doc.get("custom_machine_installation")
	if not mi:
		return
	status = frappe.db.get_value("Machine Installation", mi, "maintenance_status")
	doc.custom_warranty_status = status or ""


@frappe.whitelist()
def create_installation_ticket(
	machine_installation: str,
	subject: str | None = None,
	engineers: list | str | None = None,
) -> str:
	"""Create HD Ticket type Installation linked to Machine Installation."""
	if not frappe.db.exists("DocType", "HD Ticket"):
		frappe.throw(_("Helpdesk (HD Ticket) is required. Install the helpdesk app."))

	mi = frappe.get_doc("Machine Installation", machine_installation)
	if isinstance(engineers, str):
		import json

		engineers = json.loads(engineers) if engineers else []

	ensure_ticket_types()
	ticket = frappe.new_doc("HD Ticket")
	ticket.subject = subject or f"Installation — {mi.serial_no or mi.name}"
	if hasattr(ticket, "ticket_type"):
		ticket.ticket_type = "Installation"
	ticket.description = f"Installation service request for {mi.item_code} / {mi.serial_no}"
	ticket.custom_machine_installation = mi.name
	ticket.custom_visit_category = "Installation"
	ticket.custom_service_request_number = make_service_request_number()
	for row in engineers or []:
		ticket.append(
			"custom_service_personnel",
			{"employee": row.get("employee"), "role": row.get("role") or "Engineer"},
		)
	# Customer sync: set if field exists
	if mi.customer and frappe.db.has_column("HD Ticket", "customer"):
		# Helpdesk may use HD Customer link; set description context only if needed
		pass
	ticket.insert(ignore_permissions=True)
	return ticket.name

File: athru_service/public/js/hd_ticket.js
frappe.ui.form.on("HD Ticket", {
	refresh(frm) {
		frm.set_intro(__("Athru presents HD Ticket as a Service Request in the SPA."));
		if (frm.doc.custom_machine_installation) {
			frm.add_custom_button(__("Machine Hub"), () => {
				window.open(
					`/athru-service/machines/${encodeURIComponent(frm.doc.custom_machine_installation)}`,
					"_blank"
				);
			});
		}
		if (!frm.is_new()) {
			frm.add_custom_button(__("Maintenance Visit"), () => {
				frappe.call({
					method: "athru_service.api.spa.create_visit_from_ticket",
					args: { hd_ticket: frm.doc.name },
					freeze: true,
					callback(r) {
						if (r.message) frappe.set_route("Form", "Maintenance Visit", r.message);
					},
				});
			}, __("Create"));
			if (["Preventive", "Contract Visit", "PMC"].includes(frm.doc.ticket_type) ||
				["PMC", "AMC", "CAMC"].includes(frm.doc.custom_visit_category)) {
				frm.add_custom_button(__("Maintenance Schedule"), () => {
					frappe.call({
						method: "athru_service.api.spa.create_schedule_from_ticket",
						args: { hd_ticket: frm.doc.name },
						freeze: true,
						callback(r) {
							if (r.message) frappe.set_route("Form", "Maintenance Schedule", r.message);
						},
					});
				}, __("Create"));
			}
			frm.add_custom_button(__("Service Report"), () => {
				frappe.new_doc("Service Report", {
					hd_ticket: frm.doc.name,
					machine_installation: frm.doc.custom_machine_installation,
				});
			}, __("Create"));
		}
	},
	custom_machine_installation(frm) {
		if (!frm.doc.custom_machine_installation) return;
		frappe.db
			.get_value("Machine Installation", frm.doc.custom_machine_installation, [
				"serial_no",
				"item_code",
				"customer",
				"project",
				"current_service_contract",
				"maintenance_status",
			])
			.then((r) => {
				if (!r.message) return;
				frm.set_value("custom_serial_no", r.message.serial_no);
				frm.set_value("custom_item_code", r.message.item_code);
				frm.set_value("custom_erp_customer", r.message.customer);
				frm.set_value("custom_project", r.message.project);
				frm.set_value("custom_service_contract", r.message.current_service_contract);
				frm.set_value("custom_warranty_status", r.message.maintenance_status);
			});
	},
});
